### ISSUE TITLE
Upgrade: Create courses for Category Question banks

### ISSUE THREAD

**Issue by mattporritt** [catalyst/moodle-MDL-70329#90]
As part of the upgrade process from a pre 4.0 Moodle to a 4.0 Moodle Category Question banks with questions will need to be converted into new courses.

We need to:
* [x] Query the database to find out categories that have question banks that contain questions
* [x] Create a new course in the same category. This course will eventually contain the question bank

There is initial SQL code in issue #1 that should help identify the categories with questions.
We also need to include the system level question bank.  A decision will need to be made where this course is located.

**Comment by mattporritt** [comment]
@marcghaly the docs for the upgrade API may help you get started: https://docs.moodle.org/dev/Upgrade_API

**Comment by abonaccorso** [comment]
probably you already have thought about this @mattporritt and @marcghaly: we also need to decide who will have access to these new courses with existing question banks. As a first guess I would say: every teacher and manager in the course where the question bank comes from.

**Comment by mattporritt** [comment]
Hi @abonaccorso we have had an initial think about it. We need to confirm but believe we won't actually have to do anything explicit here with roles or enrolments in these new courses. As users who have access to the existing category level questions banks should already have the permissions required in any newly created courses in the same category. 

But we will confirm

**Comment by abonaccorso** [comment]
Rethinking it, it seems to me I made an error in my original thinking @mattporritt. Sorry for that.

**Comment by guillogo** [comment]
Hi @marcghaly 

I'll add here some information that could help in the upgrade tickets:

1. You can find the related course or category using the `contextid` 
2. That `contextid` points to a record in `mdl_context`
3. There you can find the `instanceid` (module id, course id or category id)... and how you know if is a category or a course?: with the contextlevel field

**contextlevel:**
```
define('CONTEXT_SYSTEM', 10);
define('CONTEXT_USER', 30);
define('CONTEXT_COURSECAT', 40);
define('CONTEXT_COURSE', 50);
define('CONTEXT_MODULE', 70);
define('CONTEXT_BLOCK', 80);
```

Also, can you follow the code style documentation: https://docs.moodle.org/dev/Coding_style and please let me know if you have questions.

Thank you.

**Comment by guillogo** [comment]
Hi @marcghaly 

Before moving this ticket to -> In Review, can you please check (like a pre-review):

1. That all SQL queries follow the code style https://docs.moodle.org/dev/SQL_coding_style
2. use {table} instead of mdl_table
3. Verify that your PR does this: Create a course for each category that has questions categories populated (we don't need to create a course for system and course context as we already have course for those, probably a validation will be need it).
4. re-base against master_MDL-71378

Thank you.
**Guillermo**

**Comment by marcghaly** [comment]
Hi, this is the PR for this issue. [](https://github.com/catalyst/moodle-MDL-70329/pull/113)

**Comment by guillogo** [comment]
I'm OK with merge the PR. As I worked with Marc on this one, @safatshahin or @mattporritt do you want to take a look?

**Comment by mattporritt** [comment]
Hi @marcghaly and @guillogo 
What's left to do on this one?

**Comment by guillogo** [comment]
Hi @mattporritt the changes after the last review were done in the PR: https://github.com/catalyst/moodle-MDL-70329/pull/113 Everything is done for issue 90. Thank you.

**Comment by mattporritt** [comment]
@guillogo I've merged in PR#113. Just confirming this issue is good to close now?

**Comment by guillogo** [comment]
Hi @mattporritt correct, this can be closed now.
